From fdae3d2c3cd6aafe0e3efa4b69b4527be1320adc Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@freescale.com>
Date: Sat, 29 Nov 2014 19:23:40 +0800
Subject: [PATCH 0976/1074] MLK-9932-2 usb: gadget: g_serial: fix oops during
 process context

We meet below oops at rare cases when we do g_serial test at SMP platform,
in our cases, the g_serial will be removed after some access.
It seems the tty still tries to write when the port is freed.

Unable to handle kernel NULL pointer dereference at virtual address 000000b0
pgd = 80004000
[000000b0] *pgd=00000000
Internal error: Oops: 17 1 PREEMPT SMP ARM
Modules linked in: usb_f_acm u_serial g_serial libcomposite configfs ov5642_camera ov5640_camera_mipi ov5640_camera mxc_dcic mxc_v4l2_capture ipu_bg_overlay_sdc ipu_still ipu_prp_enc ipu_csi_enc ipu_fg_overlay_sdc evbug
CPU: 0 PID: 25 Comm: kworker/0:1 Not tainted 3.10.53-02612-g3ac9b04-dirty #210
Workqueue: events flush_to_ldisc
task: a80ae580 ti: a80e8000 task.ti: a80e8000
PC is at _raw_spin_lock_irqsave+0x18/0x58
LR is at _raw_spin_lock_irqsave+0x18/0x58
pc : [<8068490c>] lr : [<8068490c>] psr: 800f0093
sp : a80e9da0 ip : 00000001 fp : a8d97c00
r10: a8d97e88 r9 : a8d97c00 r8 : 0000000d
r7 : 0000006c r6 : a8da7400 r5 : 200f0013 r4 : 000000b0
r3 : a80e8000 r2 : 00000000 r1 : a8d97c00 r0 : 00000001
Flags: Nzcv IRQs off FIQs on Mode SVC_32 ISA ARM Segment kernel
Control: 10c53c7d Table: 38cf404a DAC: 00000015
Process kworker/0:1 (pid: 25, stack limit = 0xa80e8238)
Stack: (0xa80e9da0 to 0xa80ea000)
9da0: 7f095a30 00000000 000000b0 7f095a44 a8da7400 a8d95d1f a8da7400 802d8484
9dc0: a80ae580 802d5ab0 a83a2500 a8d97e70 a8d97c00 a8da7400 a8d95d1f a8d95c1f
9de0: 0000006c 0000000d a8d97c00 00000000 a8d97c00 802d7010 a8044080 00000000
9e00: 00000000 80cc00c0 806ae500 00000000 00000000 00000002 00000000 815977d0
9e20: 00000000 00000000 00000020 00000000 80cbcc40 a8044080 ffffbe45 00000000
9e40: 00000001 81598c40 00000000 806848b8 80009038 ffffffff 00000000 8004e7c8
9e60: 80cbe000 00000002 a80e9e8c 80684b9c 00000000 80051a10 a805c780 80cbcc40
9e80: a805c780 80cbcc40 a80e9f2c 80683850 81598840 8068490c 00000000 80cceed4
9ea0: 00000000 8004e744 81598840 a8ae5400 a8da7400 a8ae5410 a93a7940 a8ae5464
9ec0: a8d95d1c 0000000f a8d95c1c 802db1f4 802db118 a8105980 a8ae5400 81598840
9ee0: 8159ba00 00000000 00000000 a80e8038 81598840 800436d4 a8105998 a80e8000
9f00: a80e8030 00000001 a80e8000 a8105980 81598854 a8105998 a80e8000 a80e8030
9f20: 00000001 a80e8000 81598840 80044358 80044220 00000000 00000000 80d15701
9f40: a80e9f64 a808fe98 00000000 a8105980 80044220 00000000 00000000 00000000
9f60: 00000000 80049730 fcbb7bfd 00000000 ffdfdbfb a8105980 00000000 00000000
9f80: a80e9f80 a80e9f80 00000000 00000000 a80e9f90 a80e9f90 a80e9fac a808fe98
9fa0: 8004967c 00000000 00000000 8000e118 00000000 00000000 00000000 00000000
9fc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
9fe0: 00000000 00000000 00000000 00000000 00000013 00000000 32fdff6f 9355fcea
[<8068490c>] (_raw_spin_lock_irqsave+0x18/0x58) from [<7f095a44>] (gs_write_room+0x14/0x60 [u_serial])
[<7f095a44>] (gs_write_room+0x14/0x60 [u_serial]) from [<802d8484>] (tty_write_room+0x18/0x24)
[<802d8484>] (tty_write_room+0x18/0x24) from [<802d5ab0>] (process_echoes+0x48/0x2a8)
[<802d5ab0>] (process_echoes+0x48/0x2a8) from [<802d7010>] (n_tty_receive_buf+0x384/0x101c)
[<802d7010>] (n_tty_receive_buf+0x384/0x101c) from [<802db1f4>] (flush_to_ldisc+0xdc/0x140)
[<802db1f4>] (flush_to_ldisc+0xdc/0x140) from [<800436d4>] (process_one_work+0xf8/0x360)
[<800436d4>] (process_one_work+0xf8/0x360) from [<80044358>] (worker_thread+0x138/0x3d4)
[<80044358>] (worker_thread+0x138/0x3d4) from [<80049730>] (kthread+0xb4/0xb8)
[<80049730>] (kthread+0xb4/0xb8) from [<8000e118>] (ret_from_fork+0x14/0x3c)
Code: e10f5000 f10c0080 e3a00001 ebe735b3 (e1943f9f)
--[ end trace b1a6974fbb8c89dc ]--

Signed-off-by: Peter Chen <peter.chen@freescale.com>
[Original patch taken from git://git.freescale.com/imx/linux-2.6-imx.git]
Signed-off-by: Biyao Zhai <biyao.zhai@windriver.com>
---
 drivers/usb/gadget/u_serial.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/drivers/usb/gadget/u_serial.c b/drivers/usb/gadget/u_serial.c
index 7c29dd8..cfbd501 100644
--- a/drivers/usb/gadget/u_serial.c
+++ b/drivers/usb/gadget/u_serial.c
@@ -946,6 +946,9 @@ static int gs_write_room(struct tty_struct *tty)
 	unsigned long	flags;
 	int		room = 0;
 
+	if (!port)
+		return room;
+
 	spin_lock_irqsave(&port->port_lock, flags);
 	if (port->port_usb)
 		room = gs_buf_space_avail(&port->port_write_buf);
-- 
1.7.5.4

